File: src/im_agent/harborbeacon.py
from __future__ import annotations

import hashlib
import json
import os
from dataclasses import dataclass, field
from typing import Any
from urllib import error, parse, request

from im_agent.models import InboundMessage

DEFAULT_CONTRACT_VERSION = "2.0"
DEFAULT_AUTONOMY_LEVEL = "supervised"
DEFAULT_INTENT_DOMAIN = "general"
DEFAULT_INTENT_ACTION = "message"
DEFAULT_SOURCE_SURFACE = "harborgate"
DEFAULT_TIMEOUT_SECONDS = 15
DEFAULT_ADMIN_TIMEOUT_SECONDS = 10
TASK_ARGS_KEY = "ar" + "gs"


def _env(primary: str) -> str:
    return os.getenv(primary, "").strip()


def _int_env(primary: str, default: int) -> int:
    raw = _env(primary)
    if not raw:
        return default
    return int(raw)


def _strip_endpoint_suffix(base_url: str, endpoint_suffix: str) -> str:
    normalized = base_url.strip()
    if not normalized:
        return ""
    if normalized.endswith(endpoint_suffix):
        return normalized[: -len(endpoint_suffix)].rstrip("/")
    suffixed = f"{endpoint_suffix}/"
    if normalized.endswith(suffixed):
        return normalized[: -len(suffixed)].rstrip("/")
    return normalized


def _canonical_json(payload: Any) -> str:
    try:
        return json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    except TypeError:
        return json.dumps(str(payload), ensure_ascii=False)


def _stable_id(prefix: str, payload: str, length: int = 24) -> str:
    digest = hashlib.sha256(payload.encode("utf-8")).hexdigest()
    return f"{prefix}{digest[:length]}"


def derive_route_key(incoming: InboundMessage) -> str:
    if incoming.route_key.strip():
        return incoming.route_key.strip()
    return _stable_id("gw_route_", f"{incoming.platform}|{incoming.chat_id}", length=20)


def derive_session_id(incoming: InboundMessage) -> str:
    if incoming.session_id.strip():
        return incoming.session_id.strip()
    return _stable_id("gw_sess_", f"{incoming.platform}|{incoming.chat_id}|{incoming.user_id}", length=20)


def _extract_dict(value: Any) -> dict[str, Any]:
    return dict(value) if isinstance(value, dict) else {}


def _extract_list(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    items: list[dict[str, Any]] = []
    for item in value:
        if isinstance(item, dict):
            items.append(dict(item))
    return items


def _event_fingerprint(incoming: InboundMessage) -> str:
    message_id = incoming.message_id.strip()
    if message_id:
        return f"{incoming.platform}|{incoming.chat_id}|{message_id}"

    raw_payload = incoming.raw_payload or {}
    candidate = (
        raw_payload.get("message_id")
        or raw_payload.get("msg_id")
        or raw_payload.get("event_id")
        or raw_payload.get("client_id")
    )
    if candidate:
        return f"{incoming.platform}|{incoming.chat_id}|{candidate}"

    payload_fingerprint = _canonical_json(
        {
            "platform": incoming.platform,
            "chat_id": incoming.chat_id,
            "user_id": incoming.user_id,
            "text": incoming.text,
            "timestamp": incoming.timestamp,
            "raw_payload": raw_payload,
        }
    )
    return f"{incoming.platform}|{incoming.chat_id}|{payload_fingerprint}"


def _intent_block(
    incoming: InboundMessage,
    *,
    default_domain: str,
    default_action: str,
) -> dict[str, str]:
    raw_payload = incoming.raw_payload or {}
    payload_intent = raw_payload.get("intent")
    payload_intent = payload_intent if isinstance(payload_intent, dict) else {}

    domain = str(
        payload_intent.get("domain")
        or raw_payload.get("domain")
        or incoming.metadata.get("domain")
        or default_domain
    ).strip() or default_domain
    action = str(
        payload_intent.get("action")
        or raw_payload.get("action")
        or incoming.metadata.get("action")
        or default_action
    ).strip() or default_action
    return {
        "domain": domain,
        "action": action,
        "raw_text": incoming.text,
    }


def build_turn_request(
    incoming: InboundMessage,
    *,
    conversation_handle: str | None = None,
    continuation: dict[str, Any] | None = None,
    autonomy_level: str = DEFAULT_AUTONOMY_LEVEL,
    default_domain: str = DEFAULT_INTENT_DOMAIN,
    default_action: str = DEFAULT_INTENT_ACTION,
    source_surface: str = DEFAULT_SOURCE_SURFACE,
) -> dict[str, Any]:
    event_fingerprint = _event_fingerprint(incoming)
    route_key = derive_route_key(incoming)
    raw_payload = incoming.raw_payload or {}
    metadata: dict[str, Any] = {}
    intent = _intent_block(
        incoming,
        default_domain=default_domain,
        default_action=default_action,
    )
    if intent:
        metadata["intent"] = intent
    entity_refs = _extract_dict(raw_payload.get("entity_refs"))
    if entity_refs:
        metadata["entity_refs"] = entity_refs
    task_args = _extract_dict(raw_payload.get(TASK_ARGS_KEY))
    if task_args:
        metadata[TASK_ARGS_KEY] = task_args

    continuation_payload = continuation if isinstance(continuation, dict) else None

    request_payload = {
        "turn": {
            "turn_id": _stable_id("turn_", event_fingerprint),
            "trace_id": _stable_id("trace_", f"trace|{event_fingerprint}"),
            "occurred_at": incoming.timestamp,
            "retry_of": None,
        },
        "actor": {
            "user_id": incoming.user_id,
            "workspace_id": str(raw_payload.get("workspace_id") or incoming.metadata.get("workspace_id") or "home-1"),
            "account_id": str(raw_payload.get("account_id") or incoming.metadata.get("account_id") or "").strip() or None,
        },
        "conversation": {
            "handle": str(conversation_handle or "").strip() or None,
            "channel": incoming.platform,
            "surface": source_surface,
            "thread_id": incoming.chat_id,
            "chat_type": incoming.chat_type.strip() or "unknown",
        },
        "transport": {
            "route_key": route_key,
            "message_id": incoming.message_id.strip(),
            "capabilities": {
                "text": True,
                "image": True,
                "file": True,
                "video": True,
            },
            "metadata": metadata,
        },
        "input": {
            "text": incoming.text,
            "parts": _extract_list(incoming.attachments),
        },
        "continuation": continuation_payload,
        "autonomy": {
            "level": autonomy_level,
        },
    }
    return request_payload


def _continuation_from_active_frame(
    active_frame: dict[str, Any] | None,
    turn_block: dict[str, Any],
) -> dict[str, Any] | None:
    if not isinstance(active_frame, dict):
        return None
    token = str(active_frame.get("continuation_token") or "").strip()
    if not token:
        return None
    return {
        "token": token,
        "frame_id": str(active_frame.get("frame_id") or "").strip(),
        "reply_to_turn_id": str(turn_block.get("turn_id") or "").strip(),
        "expires_at": active_frame.get("expires_at"),
    }


@dataclass(slots=True)
class TaskTurnResult:
    text: str
    task_id: str
    trace_id: str
    status: str
    route_key: str
    conversation_handle: str | None = None
    continuation: dict[str, Any] | None = None
    active_frame: dict[str, Any] | None = None
    delivery_hints: list[dict[str, Any]] = field(default_factory=list)
    prompt: str | None = None
    next_actions: list[str] = field(default_factory=list)
    request_payload: dict[str, Any] = field(default_factory=dict)
    response_payload: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True)
class HarborBeaconTaskClient:
    base_url: str
    api_token: str = ""
    contract_version: str = DEFAULT_CONTRACT_VERSION
    autonomy_level: str = DEFAULT_AUTONOMY_LEVEL
    default_domain: str = DEFAULT_INTENT_DOMAIN
    default_action: str = DEFAULT_INTENT_ACTION
    source_surface: str = DEFAULT_SOURCE_SURFACE
    timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS

    def submit_turn(self, incoming: InboundMessage, *, session_metadata: dict[str, object] | None = None) -> TaskTurnResult:
        metadata = session_metadata or {}
        conversation_handle = str(metadata.get("conversation_handle") or "").strip() or None
        continuation = metadata.get("continuation")
        continuation_payload = dict(continuation) if isinstance(continuation, dict) else None
        request_payload = build_turn_request(
            incoming,
            conversation_handle=conversation_handle,
            continuation=continuation_payload,
            autonomy_level=self.autonomy_level,
            default_domain=self.default_domain,
            default_action=self.default_action,
            source_surface=self.source_surface,
        )
        response_payload = self._post_json("/api/turns", request_payload)
        return self._map_turn_response(request_payload, response_payload)

    def _post_json(self, endpoint: str, payload: dict[str, Any]) -> dict[str, Any]:
        body = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        req = request.Request(
            self._url(endpoint),
            data=body,
            headers=self._headers(body),
            method="POST",
        )
        try:
            with request.urlopen(req, timeout=self.timeout_seconds) as response:
                raw = response.read().decode("utf-8")
        except error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(self._format_http_error(exc.code, detail)) from exc
        except error.URLError as exc:
            raise RuntimeError(f"Could not reach HarborBeacon task API: {exc.reason}") from exc

        data = json.loads(raw) if raw else {}
        if not isinstance(data, dict):
            raise RuntimeError("HarborBeacon task API returned a non-object JSON payload")
        return data

    def _headers(self, body: bytes) -> dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "Content-Length": str(len(body)),
            "X-Contract-Version": self.contract_version,
        }
        if self.api_token:
            headers["Authorization"] = f"Bearer {self.api_token}"
        return headers

    def _url(self, endpoint: str) -> str:
        if self.base_url.endswith(endpoint):
            return self.base_url
        return parse.urljoin(f"{self.base_url.rstrip('/')}/", endpoint.lstrip("/"))

    @staticmethod
    def _format_http_error(status_code: int, detail: str) -> str:
        message = f"HarborBeacon task API returned HTTP {status_code}"
        try:
            payload = json.loads(detail)
        except json.JSONDecodeError:
            payload = None

        if isinstance(payload, dict):
            error_block = payload.get("error")
            if isinstance(error_block, dict):
                code = str(error_block.get("code") or "").strip()
                text = str(error_block.get("message") or "").strip()
                details = " ".join(part for part in (code, text) if part).strip()
                if details:
                    return f"{message}: {details}"
        detail = detail.strip()
        return f"{message}: {detail}" if detail else message

    @staticmethod
    def _map_turn_response(request_payload: dict[str, Any], response_payload: dict[str, Any]) -> TaskTurnResult:
        turn_block = response_payload.get("turn")
        turn_block = turn_block if isinstance(turn_block, dict) else {}
        conversation_block = response_payload.get("conversation")
        conversation_block = conversation_block if isinstance(conversation_block, dict) else {}
        reply_block = response_payload.get("reply")
        reply_block = reply_block if isinstance(reply_block, dict) else {}
        active_frame = response_payload.get("active_frame")
        active_frame = active_frame if isinstance(active_frame, dict) else None
        status = str(turn_block.get("status") or "completed").strip() or "completed"
        error_block = response_payload.get("error")
        error_block = error_block if isinstance(error_block, dict) else {}
        message = str(reply_block.get("text") or "").strip()
        error_message = str(error_block.get("message") or "").strip()
        reply_text = message or error_message or "HarborBeacon returned an empty reply."
        expected_reply = active_frame.get("expected_reply") if isinstance(active_frame, dict) else []
        if not isinstance(expected_reply, list):
            expected_reply = []
        hints = response_payload.get("delivery_hints")
        hints = [dict(item) for item in hints if isinstance(item, dict)] if isinstance(hints, list) else []
        continuation = _continuation_from_active_frame(active_frame, turn_block)

        return TaskTurnResult(
            text=reply_text,
            task_id=str(turn_block.get("turn_id") or request_payload["turn"]["turn_id"]).strip(),
            trace_id=str(turn_block.get("trace_id") or request_payload["turn"]["trace_id"]).strip(),
            status=status,
            route_key=str(request_payload["transport"]["route_key"]).strip(),
            conversation_handle=str(conversation_block.get("handle") or "").strip() or None,
            continuation=continuation,
            active_frame=active_frame,
            delivery_hints=hints,
            prompt=reply_text if active_frame else None,
            next_actions=[str(item) for item in expected_reply if str(item).strip()],
            request_payload=request_payload,
            response_payload=response_payload,
        )


@dataclass(slots=True)
class HarborBeaconAdminClient:
    base_url: str
    api_token: str = ""
    contract_version: str = DEFAULT_CONTRACT_VERSION
    timeout_seconds: int = DEFAULT_ADMIN_TIMEOUT_SECONDS

    def upsert_notification_target(
        self,
        *,
        label: str,
        route_key: str,
        platform_hint: str,
        is_default: bool = False,
        target_id: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "label": label,
            "route_key": route_key,
            "platform_hint": platform_hint,
        }
        if is_default:
            payload["is_default"] = True
        if target_id and target_id.strip():
            payload["target_id"] = target_id.strip()
        return self._post_json("/api/admin/notification-targets", payload)

    def _post_json(self, endpoint: str, payload: dict[str, Any]) -> dict[str, Any]:
        body = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        req = request.Request(
            self._url(endpoint),
            data=body,
            headers=self._headers(body),
            method="POST",
        )
        try:
            with request.urlopen(req, timeout=self.timeout_seconds) as response:
                raw = response.read().decode("utf-8")
        except error.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(self._format_http_error(exc.code, detail)) from exc
        except error.URLError as exc:
            raise RuntimeError(f"Could not reach HarborBeacon admin API: {exc.reason}") from exc

        data = json.loads(raw) if raw else {}
        if not isinstance(data, dict):
            raise RuntimeError("HarborBeacon admin API returned a non-object JSON payload")
        return data

    def _headers(self, body: bytes) -> dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "Content-Length": str(len(body)),
            "X-Contract-Version": self.contract_version,
        }
        if self.api_token:
            headers["Authorization"] = f"Bearer {self.api_token}"
        return headers

    def _url(self, endpoint: str) -> str:
        if self.base_url.endswith(endpoint):
            return self.base_url
        return parse.urljoin(f"{self.base_url.rstrip('/')}/", endpoint.lstrip("/"))

    @staticmethod
    def _format_http_error(status_code: int, detail: str) -> str:
        message = f"HarborBeacon admin API returned HTTP {status_code}"
        try:
            payload = json.loads(detail)
        except json.JSONDecodeError:
            payload = None

        if isinstance(payload, dict):
            error_block = payload.get("error")
            if isinstance(error_block, dict):
                code = str(error_block.get("code") or "").strip()
                text = str(error_block.get("message") or "").strip()
                details = " ".join(part for part in (code, text) if part).strip()
                if details:
                    return f"{message}: {details}"
        detail = detail.strip()
        return f"{message}: {detail}" if detail else message


def build_harborbeacon_client_from_env() -> HarborBeaconTaskClient | None:
    base_url = _env("HARBORBEACON_TASK_API_URL")
    base_url = _strip_endpoint_suffix(base_url, "/api/turns")
    if not base_url:
        return None
    return HarborBeaconTaskClient(
        base_url=base_url,
        api_token=_env("HARBORBEACON_TASK_API_TOKEN"),
        contract_version=_env("HARBORBEACON_CONTRACT_VERSION") or DEFAULT_CONTRACT_VERSION,
        autonomy_level=_env("HARBORBEACON_AUTONOMY_LEVEL") or DEFAULT_AUTONOMY_LEVEL,
        default_domain=_env("HARBORBEACON_DEFAULT_DOMAIN") or DEFAULT_INTENT_DOMAIN,
        default_action=_env("HARBORBEACON_DEFAULT_ACTION") or DEFAULT_INTENT_ACTION,
        source_surface=_env("HARBORBEACON_SOURCE_SURFACE") or DEFAULT_SOURCE_SURFACE,
        timeout_seconds=_int_env(
            "HARBORBEACON_TASK_API_TIMEOUT_SECONDS",
            DEFAULT_TIMEOUT_SECONDS,
        ),
    )


def build_harborbeacon_admin_client_from_env() -> HarborBeaconAdminClient | None:
    base_url = _env("HARBORBEACON_ADMIN_API_URL") or _env("HARBORBEACON_TASK_API_URL")
    base_url = _strip_endpoint_suffix(base_url, "/api/turns")
    if not base_url:
        return None
    return HarborBeaconAdminClient(
        base_url=base_url,
        api_token=(
            _env("HARBORBEACON_ADMIN_API_TOKEN")
            or _env("HARBORBEACON_TASK_API_TOKEN")
            or _env("IM_AGENT_SERVICE_TOKEN")
        ),
        contract_version=_env("HARBORBEACON_CONTRACT_VERSION") or DEFAULT_CONTRACT_VERSION,
        timeout_seconds=_int_env(
            "HARBORBEACON_ADMIN_API_TIMEOUT_SECONDS",
            DEFAULT_ADMIN_TIMEOUT_SECONDS,
        ),
    )

File: rust/harborgate/src/harborbeacon.rs
use crate::config::AppConfig;
use crate::error::GatewayError;
use crate::models::InboundMessage;
use axum::http::StatusCode;
use reqwest::Client;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};

pub const DEFAULT_CONTRACT_VERSION: &str = "2.0";
pub const DEFAULT_TURN_ENDPOINT: &str = "/api/web/turns";

#[derive(Clone)]
pub struct HarborBeaconTaskClient {
    base_url: String,
    api_token: String,
    turn_endpoint: String,
    contract_version: String,
    http: Client,
}

#[derive(Debug, Clone)]
pub struct TaskTurnResult {
    pub text: String,
    pub task_id: String,
    pub trace_id: String,
    pub status: String,
    pub route_key: String,
    pub conversation_handle: Option<String>,
    pub continuation: Option<Value>,
    pub active_frame: Option<Value>,
    pub next_actions: Vec<String>,
    pub response_payload: Value,
}

impl HarborBeaconTaskClient {
    pub fn from_config(config: &AppConfig) -> Option<Self> {
        config.harborbeacon_enabled().then(|| Self {
            base_url: config
                .harborbeacon_base_url
                .trim_end_matches('/')
                .to_string(),
            api_token: config.harborbeacon_token.clone(),
            turn_endpoint: config.harborbeacon_turn_endpoint.clone(),
            contract_version: DEFAULT_CONTRACT_VERSION.to_string(),
            http: Client::new(),
        })
    }

    pub async fn submit_turn(
        &self,
        incoming: &InboundMessage,
        session_metadata: &serde_json::Map<String, Value>,
    ) -> Result<TaskTurnResult, GatewayError> {
        let conversation_handle = session_metadata
            .get("conversation_handle")
            .and_then(Value::as_str)
            .filter(|value| !value.trim().is_empty())
            .map(str::to_string);
        let continuation = session_metadata
            .get("continuation")
            .filter(|value| value.is_object())
            .cloned();
        let request_payload =
            build_turn_request(incoming, conversation_handle.as_deref(), continuation);
        let response_payload = self.post_json(&request_payload).await?;
        Ok(map_turn_response(&request_payload, response_payload))
    }

    async fn post_json(&self, payload: &Value) -> Result<Value, GatewayError> {
        let url = format!(
            "{}/{}",
            self.base_url.trim_end_matches('/'),
            self.turn_endpoint.trim_start_matches('/')
        );
        let mut request = self
            .http
            .post(url)
            .header("X-Contract-Version", &self.contract_version)
            .json(payload);
        if !self.api_token.trim().is_empty() {
            request = request.bearer_auth(&self.api_token);
        }
        let response = request.send().await.map_err(|err| {
            GatewayError::infrastructure(format!("Could not reach HarborBeacon task API: {err}"))
        })?;
        let status = response.status();
        let body = response.text().await.map_err(|err| {
            GatewayError::infrastructure(format!(
                "Could not read HarborBeacon task API response: {err}"
            ))
        })?;
        let payload: Value = if body.trim().is_empty() {
            json!({})
        } else {
            serde_json::from_str(&body).map_err(|err| {
                GatewayError::infrastructure(format!(
                    "HarborBeacon task API returned invalid JSON: {err}"
                ))
            })?
        };
        if !status.is_success() {
            let message = payload
                .get("error")
                .and_then(Value::as_object)
                .and_then(|error| {
                    let code = error.get("code").and_then(Value::as_str).unwrap_or("");
                    let text = error.get("message").and_then(Value::as_str).unwrap_or("");
                    let detail = [code, text]
                        .into_iter()
                        .filter(|part| !part.is_empty())
                        .collect::<Vec<_>>()
                        .join(" ");
                    (!detail.is_empty()).then_some(detail)
                })
                .unwrap_or_else(|| body.trim().to_string());
            return Err(GatewayError::new(
                StatusCode::BAD_GATEWAY,
                "UPSTREAM_TASK_API_ERROR",
                format!("HarborBeacon task API returned HTTP {status}: {message}"),
            ));
        }
        Ok(payload)
    }
}

pub fn derive_route_key(incoming: &InboundMessage) -> String {
    if !incoming.route_key.trim().is_empty() {
        return incoming.route_key.trim().to_string();
    }
    stable_id(
        "gw_route_",
        &format!("{}|{}", incoming.platform, incoming.chat_id),
        20,
    )
}

pub fn derive_session_id(incoming: &InboundMessage) -> String {
    if !incoming.session_id.trim().is_empty() {
        return incoming.session_id.trim().to_string();
    }
    stable_id(
        "gw_sess_",
        &format!(
            "{}|{}|{}",
            incoming.platform, incoming.chat_id, incoming.user_id
        ),
        20,
    )
}

pub fn build_turn_request(
    incoming: &InboundMessage,
    conversation_handle: Option<&str>,
    continuation: Option<Value>,
) -> Value {
    let event_fingerprint = event_fingerprint(incoming);
    let route_key = derive_route_key(incoming);
    let raw_payload = incoming.raw_payload.as_object();
    let mut metadata = serde_json::Map::new();
    metadata.insert(
        "intent".into(),
        json!({
            "domain": raw_lookup(raw_payload, "domain").or_else(|| metadata_lookup(&incoming.metadata, "domain")).unwrap_or_else(|| "general".to_string()),
            "action": raw_lookup(raw_payload, "action").or_else(|| metadata_lookup(&incoming.metadata, "action")).unwrap_or_else(|| "message".to_string()),
            "raw_text": incoming.text,
        }),
    );
    if let Some(entity_refs) = raw_payload
        .and_then(|raw| raw.get("entity_refs"))
        .filter(|value| value.is_object())
    {
        metadata.insert("entity_refs".into(), entity_refs.clone());
    }
    if let Some(args) = raw_payload
        .and_then(|raw| raw.get("args"))
        .filter(|value| value.is_object())
    {
        metadata.insert("args".into(), args.clone());
    }

    json!({
        "turn": {
            "turn_id": stable_id("turn_", &event_fingerprint, 24),
            "trace_id": stable_id("trace_", &format!("trace|{event_fingerprint}"), 24),
            "occurred_at": incoming.timestamp,
            "retry_of": null,
        },
        "actor": {
            "user_id": incoming.user_id,
            "workspace_id": raw_lookup(raw_payload, "workspace_id").or_else(|| metadata_lookup(&incoming.metadata, "workspace_id")).unwrap_or_else(|| "home-1".to_string()),
            "account_id": raw_lookup(raw_payload, "account_id").or_else(|| metadata_lookup(&incoming.metadata, "account_id")),
        },
        "conversation": {
            "handle": conversation_handle.filter(|value| !value.trim().is_empty()),
            "channel": incoming.platform,
            "surface": "harborgate",
            "thread_id": incoming.chat_id,
            "chat_type": if incoming.chat_type.trim().is_empty() { "unknown" } else { incoming.chat_type.as_str() },
        },
        "transport": {
            "route_key": route_key,
            "message_id": incoming.message_id.trim(),
            "capabilities": {
                "text": true,
                "image": true,
                "file": true,
                "video": true,
            },
            "metadata": metadata,
        },
        "input": {
            "text": incoming.text,
            "parts": incoming.attachments,
        },
        "continuation": continuation,
        "autonomy": {
            "level": "supervised",
        },
    })
}

fn map_turn_response(request_payload: &Value, response_payload: Value) -> TaskTurnResult {
    let turn = response_payload.get("turn").and_then(Value::as_object);
    let conversation = response_payload
        .get("conversation")
        .and_then(Value::as_object);
    let reply = response_payload.get("reply").and_then(Value::as_object);
    let active_frame = response_payload
        .get("active_frame")
        .filter(|value| value.is_object())
        .cloned();
    let error = response_payload.get("error").and_then(Value::as_object);
    let text = reply
        .and_then(|reply| reply.get("text"))
        .and_then(Value::as_str)
        .or_else(|| {
            error
                .and_then(|error| error.get("message"))
                .and_then(Value::as_str)
        })
        .unwrap_or("HarborBeacon returned an empty reply.")
        .trim()
        .to_string();
    let task_id = turn
        .and_then(|turn| turn.get("turn_id"))
        .and_then(Value::as_str)
        .or_else(|| {
            request_payload
                .pointer("/turn/turn_id")
                .and_then(Value::as_str)
        })
        .unwrap_or("")
        .to_string();
    let trace_id = turn
        .and_then(|turn| turn.get("trace_id"))
        .and_then(Value::as_str)
        .or_else(|| {
            request_payload
                .pointer("/turn/trace_id")
                .and_then(Value::as_str)
        })
        .unwrap_or("")
        .to_string();
    let continuation = continuation_from_active_frame(active_frame.as_ref(), turn);
    let next_actions = active_frame
        .as_ref()
        .and_then(|frame| frame.get("expected_reply"))
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_str)
                .filter(|item| !item.trim().is_empty())
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default();

    TaskTurnResult {
        text,
        task_id,
        trace_id,
        status: turn
            .and_then(|turn| turn.get("status"))
            .and_then(Value::as_str)
            .unwrap_or("completed")
            .to_string(),
        route_key: request_payload
            .pointer("/transport/route_key")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        conversation_handle: conversation
            .and_then(|conversation| conversation.get("handle"))
            .and_then(Value::as_str)
            .filter(|value| !value.trim().is_empty())
            .map(str::to_string),
        continuation,
        active_frame,
        next_actions,
        response_payload,
    }
}

fn continuation_from_active_frame(
    active_frame: Option<&Value>,
    turn: Option<&serde_json::Map<String, Value>>,
) -> Option<Value> {
    let frame = active_frame?.as_object()?;
    let token = frame.get("continuation_token")?.as_str()?.trim();
    if token.is_empty() {
        return None;
    }
    Some(json!({
        "token": token,
        "frame_id": frame.get("frame_id").and_then(Value::as_str).unwrap_or(""),
        "reply_to_turn_id": turn.and_then(|turn| turn.get("turn_id")).and_then(Value::as_str).unwrap_or(""),
        "expires_at": frame.get("expires_at").cloned().unwrap_or(Value::Null),
    }))
}

fn event_fingerprint(incoming: &InboundMessage) -> String {
    if !incoming.message_id.trim().is_empty() {
        return format!(
            "{}|{}|{}",
            incoming.platform, incoming.chat_id, incoming.message_id
        );
    }
    for key in ["message_id", "msg_id", "event_id", "client_id"] {
        if let Some(value) = incoming.raw_payload.get(key).and_then(Value::as_str) {
            if !value.trim().is_empty() {
                return format!("{}|{}|{}", incoming.platform, incoming.chat_id, value);
            }
        }
    }
    format!(
        "{}|{}|{}",
        incoming.platform,
        incoming.chat_id,
        canonical_json(&json!({
            "platform": incoming.platform,
            "chat_id": incoming.chat_id,
            "user_id": incoming.user_id,
            "text": incoming.text,
            "timestamp": incoming.timestamp,
            "raw_payload": incoming.raw_payload,
        }))
    )
}

pub fn stable_id(prefix: &str, payload: &str, length: usize) -> String {
    let mut hasher = Sha256::new();
    hasher.update(payload.as_bytes());
    let digest = format!("{:x}", hasher.finalize());
    format!("{prefix}{}", &digest[..length.min(digest.len())])
}

pub fn canonical_json(payload: &Value) -> String {
    match payload {
        Value::Object(map) => {
            let mut entries = map.iter().collect::<Vec<_>>();
            entries.sort_by(|left, right| left.0.cmp(right.0));
            let body = entries
                .into_iter()
                .map(|(key, value)| {
                    format!(
                        "{}:{}",
                        serde_json::to_string(key).unwrap(),
                        canonical_json(value)
                    )
                })
                .collect::<Vec<_>>()
                .join(",");
            format!("{{{body}}}")
        }
        Value::Array(items) => {
            let body = items
                .iter()
                .map(canonical_json)
                .collect::<Vec<_>>()
                .join(",");
            format!("[{body}]")
        }
        _ => serde_json::to_string(payload).unwrap_or_else(|_| "\"\"".to_string()),
    }
}

fn raw_lookup(raw_payload: Option<&serde_json::Map<String, Value>>, key: &str) -> Option<String> {
    raw_payload?
        .get(key)
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .map(str::to_string)
}

fn metadata_lookup(metadata: &serde_json::Map<String, Value>, key: &str) -> Option<String> {
    metadata
        .get(key)
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .map(str::to_string)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::utc_now_iso;

    #[test]
    fn build_turn_request_uses_web_contract_shape() {
        let incoming = InboundMessage {
            platform: "feishu".into(),
            chat_id: "oc_123".into(),
            user_id: "ou_123".into(),
            text: "hello".into(),
            message_id: "om_123".into(),
            chat_type: "p2p".into(),
            route_key: "".into(),
            session_id: "".into(),
            mentions: vec![],
            attachments: vec![],
            metadata: serde_json::Map::new(),
            timestamp: utc_now_iso(),
            raw_payload: json!({"message_id": "om_123"}),
        };

        let payload = build_turn_request(&incoming, Some("conv_1"), None);

        assert_eq!(payload["conversation"]["handle"], "conv_1");
        assert_eq!(payload["conversation"]["surface"], "harborgate");
        assert_eq!(
            payload["transport"]["route_key"],
            derive_route_key(&incoming)
        );
        assert!(payload.get("args").is_none());
        assert!(payload.get("source").is_none());
    }
}
